File: iOS/AUiKit/Classes/Service/Impl/AUiUserServiceImpl.swift
//
//  AUIUserServiceImpl.swift
//  AUIKit
//
//  Created by wushengtao on 2023/4/7.
//

import Foundation


open class AUIUserServiceImpl: NSObject {
    private var userList: [AUIUserInfo] = []
    private var respDelegates: NSHashTable<AnyObject> = NSHashTable<AnyObject>.weakObjects()
    private var channelName: String!
    private let rtmManager: AUIRtmManager!
    private let roomManager: AUIRoomManagerDelegate!
    
    deinit {
        aui_info("deinit AUIUserServiceImpl", tag: "AUIUserServiceImpl")
        self.rtmManager.unsubscribeUser(channelName: channelName, delegate: self)
    }
    
    public init(channelName: String, rtmManager: AUIRtmManager, roomManager: AUIRoomManagerDelegate) {
        self.rtmManager = rtmManager
        self.channelName = channelName
        self.roomManager = roomManager
        super.init()
        self.rtmManager.subscribeUser(channelName: channelName, delegate: self)
        aui_info("init AUIUserServiceImpl", tag: "AUIUserServiceImpl")
    }
}


extension AUIUserServiceImpl: AUIRtmUserProxyDelegate {
    public func onUserDidUpdated(channelName: String, userId: String, userInfo: [String : Any]) {
        aui_info("onUserDidUpdated: \(userId)", tag: "AUIUserServiceImpl")
        let user = AUIUserInfo.yy_model(withJSON: userInfo)!
        user.userId = userId
        if let oldUser = self.userList.first(where: {$0.userId == userId}) {
            if oldUser.muteAudio != user.muteAudio {
                oldUser.muteAudio = user.muteAudio
                self.respDelegates.allObjects.forEach { obj in
                    guard let obj = obj as? AUIUserRespDelegate else {return}
                    obj.onUserAudioMute(userId: userId, mute: user.muteAudio)
                }
            }
            
            if oldUser.muteVideo != user.muteVideo {
                oldUser.muteVideo = user.muteVideo
                self.respDelegates.allObjects.forEach { obj in
                    guard let obj = obj as? AUIUserRespDelegate else {return}
                    obj.onUserVideoMute(userId: userId, mute: user.muteVideo)
                }
            }
        }
        
        if let idx = self.userList.firstIndex(where: {$0.userId == userId}) {
            self.userList.replaceSubrange(idx...idx, with: [user])
        } else {
            self.userList.append(user)
        }
        self.respDelegates.allObjects.forEach { obj in
            guard let obj = obj as? AUIUserRespDelegate else {return}
            obj.onRoomUserUpdate(roomId: channelName, userInfo: user)
        }
    }
    
    public func onUserSnapshotRecv(channelName: String, userId: String, userList: [[String : Any]]) {
        aui_info("onUserSnapshotRecv: \(userId)", tag: "AUIUserServiceImpl")
        guard let users = NSArray.yy_modelArray(with: AUIUserInfo.self, json: userList) as? [AUIUserInfo] else {
            assert(false, "onUserSnapshotRecv recv fail")
            return
        }
        self.respDelegates.allObjects.forEach { obj in
            guard let obj = obj as? AUIUserRespDelegate else {return}
            self.userList = users
            obj.onRoomUserSnapshot(roomId: channelName, userList: users)
        }
        
        //对于2.1.0版本。我们推荐在join之后收到snapshot之后再去设置state
        _setupUserAttr(roomId: channelName) { error in
            //TODO: retry if fail
        }
    }
    
    public func onUserDidJoined(channelName: String, userId: String, userInfo: [String : Any]) {
        aui_info("onUserDidJoined: \(userId)", tag: "AUIUserServiceImpl")
        let user = AUIUserInfo.yy_model(withJSON: userInfo)!
        user.userId = userId
        self.userList.append(user)
        self.respDelegates.allObjects.forEach { obj in
            guard let obj = obj as? AUIUserRespDelegate else {return}
            obj.onRoomUserEnter(roomId: channelName, userInfo: user)
        }
    }
    
    public func onUserDidLeaved(channelName: String, userId: String, userInfo: [String : Any]) {
        aui_info("onUserDidLeaved: \(userId)", tag: "AUIUserServiceImpl")
        let user = userList.filter({$0.userId == userId}).first ?? AUIUserInfo.yy_model(withJSON: userInfo)!
        self.userList = userList.filter({$0.userId != userId})
        self.respDelegates.allObjects.forEach { obj in
            guard let obj = obj as? AUIUserRespDelegate else {return}
            obj.onRoomUserLeave(roomId: channelName, userInfo: user)
        }
    }
}

extension AUIUserServiceImpl: AUIUserServiceDelegate {
    public func getChannelName() -> String {
        return channelName
    }
    
    public func bindRespDelegate(delegate: AUIUserRespDelegate) {
        respDelegates.add(delegate)
    }
    
    public func unbindRespDelegate(delegate: AUIUserRespDelegate) {
        respDelegates.remove(delegate)
    }
    
    public func getUserInfoList(roomId: String, userIdList: [String], callback:@escaping AUIUserListCallback) {
        self.rtmManager.whoNow(channelName: roomId) { error, userList in
            if let error = error {
                callback(error, nil)
                return
            }
            
            var users = [AUIUserInfo]()
            userList?.forEach { attr in
                let user = AUIUserInfo.yy_model(withJSON: attr)!
                users.append(user)
            }
            self.userList = users
            callback(nil, users)
        }
    }
    
//    public func getUserInfo(by userId: String) -> AUIUserThumbnailInfo? {
//        return self.userList.filter {$0.userId == userId}.first
//    }
    
    public func muteUserAudio(isMute: Bool, callback: @escaping AUICallback) {
        aui_info("muteUserAudio: \(isMute)", tag: "AUIUserServiceImpl")
        let currentUserId = getRoomContext().currentUserInfo.userId
        let currentUser = userList.first(where: {$0.userId == currentUserId})
        currentUser?.muteAudio = isMute
        let userAttr = currentUser?.yy_modelToJSONObject() as? [String: Any] ?? [:]
//        print("muteUserAudio user:  \(userDic)")
        self.rtmManager.setPresenceState(channelName: channelName, attr: userAttr) {[weak self] error in
            guard let self = self else {return}
            if let error = error {
                callback(error)
                return
            }
            
            callback(nil)
            
            //自己状态不会更新，在这里手动回调
            self.respDelegates.allObjects.forEach { obj in
                guard let obj = obj as? AUIUserRespDelegate else {return}
                obj.onUserAudioMute(userId: currentUserId, mute: isMute)
            }
        }
    }
    
    public func muteUserVideo(isMute: Bool, callback: @escaping AUICallback) {
        aui_info("muteUserVideo: \(isMute)", tag: "AUIUserServiceImpl")
        let currentUserId = getRoomContext().currentUserInfo.userId
        let currentUser = userList.first(where: {$0.userId == currentUserId})
        currentUser?.muteVideo = isMute
        let userAttr = currentUser?.yy_modelToJSONObject() as? [String: Any] ?? [:]
//        print("muteUserAudio user:  \(userDic)")
        self.rtmManager.setPresenceState(channelName: channelName, attr: userAttr) {[weak self] error in
            guard let self = self else {return}
            if let error = error {
                callback(error)
                return
            }
            
            callback(nil)
            
            //自己状态不会更新，在这里手动回调
            self.respDelegates.allObjects.forEach { obj in
                guard let obj = obj as? AUIUserRespDelegate else {return}
                obj.onUserVideoMute(userId: currentUserId, mute: isMute)
            }
        }
    }
}

extension AUIUserServiceImpl {
    //设置用户属性到presence
    private func _setupUserAttr(roomId: String, completion: ((Error?) -> ())?) {
        let userId = AUIRoomContext.shared.currentUserInfo.userId
        let userInfo = self.userList.filter({$0.userId == userId}).first ?? AUIUserInfo()
        userInfo.userId = AUIRoomContext.shared.currentUserInfo.userId
        userInfo.userName = AUIRoomContext.shared.currentUserInfo.userName
        userInfo.userAvatar = AUIRoomContext.shared.currentUserInfo.userAvatar
        
        let userAttr = userInfo.yy_modelToJSONObject() as? [String: Any] ?? [:]
        aui_info("_setupUserAttr: \(roomId) : \(userAttr)", tag: "AUIUserServiceImpl")
        self.rtmManager.setPresenceState(channelName: roomId, attr: userAttr) { error in
            defer {
                completion?(error)
            }
            if let error = error {
                aui_info("_setupUserAttr: \(roomId) fail: \(error.localizedDescription)", tag: "AUIUserServiceImpl")
                //TODO: retry
                return
            }
            
            //rtm不会返回自己更新的数据，需要手动处理
            self.onUserDidUpdated(channelName: roomId, userId: userId, userInfo: userAttr)
        }
    }
}

File: Android/auikit/src/main/java/io/agora/auikit/service/imp/AUIUserServiceImpl.kt
package io.agora.auikit.service.imp

import io.agora.auikit.model.AUIRoomContext
import io.agora.auikit.model.AUIUserInfo
import io.agora.auikit.service.IAUIUserService
import io.agora.auikit.service.callback.AUICallback
import io.agora.auikit.service.callback.AUIException
import io.agora.auikit.service.callback.AUIUserListCallback
import io.agora.auikit.service.rtm.AUIRtmManager
import io.agora.auikit.service.rtm.AUiRtmUserProxyDelegate
import io.agora.auikit.utils.AUiLogger
import io.agora.auikit.utils.DelegateHelper
import io.agora.auikit.utils.GsonTools

class AUIUserServiceImpl constructor(
    private val channelName: String,
    private val rtmManager: AUIRtmManager
) : IAUIUserService, AUiRtmUserProxyDelegate {

    private val TAG = "AUiUserServiceImpl"

    private var mUserList = mutableListOf<AUIUserInfo>()

    private val delegateHelper = DelegateHelper<IAUIUserService.AUIUserRespDelegate>()

    init {
        rtmManager.subscribeUser(this)
    }

    override fun bindRespDelegate(delegate: IAUIUserService.AUIUserRespDelegate?) {
        delegateHelper.bindDelegate(delegate)
    }

    override fun unbindRespDelegate(delegate: IAUIUserService.AUIUserRespDelegate?) {
        delegateHelper.unBindDelegate(delegate)
    }

    override fun getUserInfoList(
        roomId: String,
        userIdList: MutableList<String>?,
        callback: AUIUserListCallback?
    ) {
        rtmManager.whoNow(roomId) { error, userList ->
            if (error != null) {
                callback?.onResult(
                    AUIException(
                        error.code,
                        error.message
                    ), null)
                return@whoNow
            }
            if (userList == null) {
                callback?.onResult(
                    AUIException(
                        -1,
                        ""
                    ), null)
                return@whoNow
            }
            val users = mutableListOf<AUIUserInfo>()
            userList.forEach { userMap ->
                GsonTools.toBean(GsonTools.beanToString(userMap), AUIUserInfo::class.java)?.let {
                    users.add(it)
                }
            }
            callback?.onResult(null, users)
        }
    }

    override fun muteUserAudio(isMute: Boolean, callback: AUICallback?) {
        val currentUserId = roomContext.currentUserInfo.userId
        val user = mUserList.first { it.userId == currentUserId }
        user.muteAudio = if (isMute) 1 else 0
        val map = GsonTools.beanToMap(user)
        rtmManager.setPresenceState(channelName, map.mapValues { it.value.toString() }) { error ->
            if (error != null) {
                callback?.onResult(
                    AUIException(
                        error.code,
                        error.message
                    )
                )
            } else {
                this.delegateHelper.notifyDelegate {
                    it.onUserAudioMute(currentUserId, isMute)
                }
                callback?.onResult(null)
            }
        }
    }

    override fun muteUserVideo(isMute: Boolean, callback: AUICallback?) {
        val currentUserId = roomContext.currentUserInfo.userId
        val user = mUserList.firstOrNull { it.userId == currentUserId }
        if (user == null) {
            callback?.onResult(
                AUIException(
                    -1,
                    "can't find current user from users"
                )
            )
            return
        }
        user.muteVideo = if (isMute) 1 else 0
        val map = GsonTools.beanToMap(user) as Map<String, String>
        rtmManager.setPresenceState(channelName, map) { error ->
            if (error != null) {
                callback?.onResult(
                    AUIException(
                        error.code,
                        error.message
                    )
                )
            } else {
                callback?.onResult(null)
            }
        }
    }

    override fun getUserInfo(userId: String): AUIUserInfo? {
        return mUserList.firstOrNull { it.userId == userId }
    }

    override fun getRoomContext(): AUIRoomContext { return AUIRoomContext.shared() }

    override fun getChannelName() = channelName

    /** AUiRtmUserProxyDelegate */
    override fun onUserSnapshotRecv(
        channelName: String,
        userId: String,
        userList: List<Map<String, Any>>
    ) {
        val users = mutableListOf<AUIUserInfo>()
        userList.forEach { userMap ->
            GsonTools.toBean(GsonTools.beanToString(userMap), AUIUserInfo::class.java)?.let {
                users.add(it)
            }
        }
        mUserList = users
        this.delegateHelper.notifyDelegate {
            it.onRoomUserSnapshot(channelName, mUserList)
        }
        setupUserAttr(channelName)
    }

    override fun onUserDidJoined(
        channelName: String,
        userId: String,
        userInfo: Map<String, Any>
    ) {
        GsonTools.toBean(GsonTools.beanToString(userInfo), AUIUserInfo::class.java)?.let { info ->
            info.userId = userId
            mUserList.add(info)
            this.delegateHelper.notifyDelegate {
                it.onRoomUserEnter(channelName, info)
            }
        }
    }

    override fun onUserDidLeaved(
        channelName: String,
        userId: String,
        userInfo: Map<String, Any>
    ) {
        val index = mUserList.indexOfFirst{ it.userId == userId }
        val info = mUserList.removeAt(index)
        this.delegateHelper.notifyDelegate {
            it.onRoomUserLeave(channelName, info)
        }
    }

    override fun onUserDidUpdated(
        channelName: String,
        userId: String,
        userInfo: Map<String, Any>
    ) {
        if (userInfo.isEmpty()) {
            return
        }
        GsonTools.toBean(GsonTools.beanToString(userInfo), AUIUserInfo::class.java)?.let { info ->
            val index = mUserList.indexOfFirst{ it.userId == info.userId }
            if (index == -1) { // 不存在该用户
                mUserList.add(info)
            } else {
                val oldInfo = mUserList[index]
                mUserList[index] = info
                // 单独更新语音被禁用回调
                if (oldInfo.muteAudio != info.muteAudio) {
                    this.delegateHelper.notifyDelegate {
                        it.onUserAudioMute(info.userId, (info.muteAudio == 1))
                    }
                }
            }
            this.delegateHelper.notifyDelegate {
                it.onRoomUserUpdate(channelName, info)
            }
        }
    }

    private fun setupUserAttr(roomId: String){
        val userId = AUIRoomContext.shared().currentUserInfo.userId
        val userInfo = mUserList.firstOrNull { it.userId == userId } ?: AUIUserInfo()
        userInfo.userId = AUIRoomContext.shared().currentUserInfo.userId
        userInfo.userName = AUIRoomContext.shared().currentUserInfo.userName
        userInfo.userAvatar = AUIRoomContext.shared().currentUserInfo.userAvatar

        val userAttr = GsonTools.beanToMap(userInfo)
        AUiLogger.logger().d(TAG, "setupUserAttr: $roomId : $userAttr")
        rtmManager.setPresenceState(roomId, userAttr) { error ->
            if(error != null){
                AUiLogger.logger().d(TAG, "setupUserAttr: $roomId fail: ${error.reason}")
            }else{
                //rtm不会返回自己更新的数据，需要手动处理
                onUserDidUpdated(roomId, userId, userAttr)
            }
        }
    }
}